### PR TITLE
Match max

### PR DESCRIPTION
**Pull request by maxTarlov**
Implement MatchMax constraints as defined in issue #30. There is also an implementation for matchMax that follows a slightly different interpretation of the definition as a result of a mistake in reading the definition set forth in issue #30.

The functions that follows the definition in issue #30 are implemented in the interface.

### CODE REVIEW

File: constraints/recursiveCatEvals.js [hunk 1]
```diff
@@ -35,29 +35,47 @@ function isMaximal(parent, child){
...
+		mytree.parentCat = parcat; //correct parentCat
+		mytree.isMax = void(0); //set to undefined
+		mytree.isMin = void(0); //set to undefined
+	}
+
 	// Check for maximalitys
-	if(!mytree.hasOwnProperty('isMax')){
-		mytree.isMax = (mytree.cat !== parcat)
+	//mytree.isMax might be undefined (void(0)) as a result of if statement
+	if(!mytree.hasOwnProperty('isMax') || mytree.isMax === void(0)){
```
Comment (jbellik) [inline constraints/recursiveCatEvals.js]: I don't think we need this if conditional at all.

File: constraints/recursiveCatEvals.js [hunk 2]
```diff
@@ -35,29 +35,47 @@ function isMaximal(parent, child){
...
+		mytree.isMax = (mytree.cat !== parcat);
 	}
-	
+
 	// Check for minimality
-	if(!mytree.hasOwnProperty('isMin')){
+	//mytree.isMin might be undefined (void(0)) as a result of if statement
+	if(!mytree.hasOwnProperty('isMin') || mytree.isMin === void(0)){
```
Comment (jbellik) [inline constraints/recursiveCatEvals.js]: Again, I think this conditional isn't necessary.

File: test/matchMaxTest.html [hunk 3]
```diff
@@ -0,0 +1,80 @@
+<!-- Max Tarlov 7/1/19 testing nonRecPairs -->
```
Comment (jbellik) [inline test/matchMaxTest.html]: I take it this comment is out-of-date.

File: constraints/recursiveCatEvals.js (hunk 1 above)
Comment (maxTarlov) [reply constraints/recursiveCatEvals.js]: once the property is initialized, hasOwnProperty() will return true even if the property is set to void(0).

File: constraints/match.js [hunk 4]
```diff
@@ -111,5 +111,93 @@ function hasMatch(sNode, pTree)
...
+		}
+
+ 		else
+ 		//the current prosodic node is non-terminal (has children)
+ 		{
+ 			for(var i = 0; i < pTree.children.length; i++)
+ 			//check each child to see if the match exists in the subtree rooted in that child
+ 			{
+ 				var child = pTree.children[i];
+ 				if(hasMatch(sNode, child)){
```
Comment (jbellik) [inline constraints/match.js]: What ensures that the match is maximal in this case?

File: constraints/recursiveCatEvals.js (hunk 1 above)
Comment (maxTarlov) [reply constraints/recursiveCatEvals.js]: issue resolved in latest commit

File: constraints/recursiveCatEvals.js (hunk 2 above)
Comment (maxTarlov) [reply constraints/recursiveCatEvals.js]: issue resolved in latest commit

File: test/matchMaxTest.html (hunk 3 above)
Comment (maxTarlov) [reply test/matchMaxTest.html]: it is. changed in latest commit

File: constraints/match.js (hunk 4 above)
Comment (maxTarlov) [reply constraints/match.js]: copy and past error. called hasMaxMatch in latest commit

**Comment by maxTarlov** [comment]
I tried to improve readability as per your request. Let me know if there are more ways I can make the code understandable.

File: constraints/match.js [hunk 5]
```diff
@@ -24,26 +24,52 @@ function getLeaves(x)
...
+ * Order insensitive version of sameIdsOrdered.
+ */
+function sameIds(a1, a2){
+	if (a1.length !== a2.length){
+		return false;
+	}
+	for (var x = 0; x < a1.length; x ++){ // for each element in a1
+		var y = 0;
+		var matched = false; // keeps track of if a1[x] has a match in a2
+		while (matched == false){//there is an element in a2 ...
```
Comment (jbellik) [inline constraints/match.js]: I'm going to give in to my impulse to nitpick here and suggest:
while(matched == false && y < a2.length)

File: constraints/match.js [hunk 6]
```diff
@@ -111,5 +137,93 @@ function hasMatch(sNode, pTree)
...
+		 }
+	 }
+	 if (sTree.cat === sCat && sTree.isMax && !hasMaxMatch(sTree, pTree)){
+		 //add violation if this node has no maximal match, is maximal and of the right cat
+		 vcount ++;
+	 }
+	 return vcount;
+ }
+
+/* Same as matchMaxSP, except matching prosodic node need not be maximal,
```
Comment (jbellik) [inline constraints/match.js]: Just for clarification, please add the following to the comment here:
Match-SP(scat-max, pcat)
ex. Match a maximal xp with any phi

File: constraints/match.js [hunk 7]
```diff
@@ -111,5 +137,93 @@ function hasMatch(sNode, pTree)
...
+ 				if(hasMaxMatch(sNode, child)){
+ 					return true;
+				}
+ 			}
+ 			return false;
+ 		}
+
+ }
+
+ /* Assign a violation for every node of syntactic category s that is not
```
Comment (jbellik) [inline constraints/match.js]: Just for clarification, please add the following to the comment here:
Match-SP(scat-max, pcat-max)
ex. Match a maximal xp with a maximal phi

File: constraints/recursiveCatEvals.js [hunk 8]
```diff
@@ -34,50 +34,49 @@ function isMaximal(parent, child){
...
-	}
-	
-	// Check for minimality
-	if(!mytree.hasOwnProperty('isMin')){
-		mytree.isMin = isMinimal(mytree);
+/* Function that takes a tree and labels all the nodes in the tree as being
+ * a minimal or maximal instance of whatever category k they are, where:
+ * minimal = does not dominate any nodes of category k
+ * maximal = is not dominated by any nodes of category k
+ * and level ordering is assumed (a node of category level k will never be
```
Comment (jbellik) [inline constraints/recursiveCatEvals.js]: Usually what you've called "level ordering" is called "layering", but I think your definition here is good.

File: constraints/recursiveCatEvals.js [hunk 9]
```diff
@@ -34,50 +34,49 @@ function isMaximal(parent, child){
...
+ * dominated by a node of category < k).
+ *
+ * This can be called in a recursive function and is compatable with GEN's
```
Comment (jbellik) [inline constraints/recursiveCatEvals.js]: Good information. Thank you for documenting this.

File: constraints/recursiveCatEvals.js [hunk 10]
```diff
@@ -34,50 +34,49 @@ function isMaximal(parent, child){
...
+	mytree.isMax = (mytree.cat !== mytree.parentCat);
+
+	//mark minimality (relies on isMinimal above)
+	mytree.isMin = isMinimal(mytree);
+
 	if(mytree.children && mytree.children.length){
-		parcat = mytree.cat;
 		for(var i = 0; i < mytree.children.length; i++){
-			mytree.children[i] = markMinMax(mytree.children[i], parcat);
+			var child = mytree.children[i];
```
Comment (jbellik) [inline constraints/recursiveCatEvals.js]: Looks good.

File: constraints/recursiveCatEvals.js (hunk 8 above)
Comment (maxTarlov) [reply constraints/recursiveCatEvals.js]: whoops, looks like I confused terms from morphology-phonology interface and syntax-prosody interface. The latest commit says "layering"

File: constraints/match.js (hunk 6 above)
Comment (maxTarlov) [reply constraints/match.js]: added in latest commit.

File: constraints/match.js (hunk 7 above)
Comment (maxTarlov) [reply constraints/match.js]: added in latest commit

**Pull request merged**